Support AES wrap and wrap_pad

mingweishih opened the pull request:
This PR adds the support of AES wrap (RFC 3394, requires OpenSSL 1.0.2) and wrap_pad (RFC 5649 requires OpenSSL 1.1.0) to `Cipher`.

OpenSSL currently requires user to explicitly opt-in the feature by setting the `EVP_CIPHER_CTX_FLAG_WRAP_ALLOW` flag in `EVP_CIPHER_CTX` before `EVP_Encryptinit_ex`/`EVP_Decryptinit_ex`. To support this mechanism, this PR also exposes the `EVP_CTX_set_flags` function that takes a bitflags value as an argument (currently only supports `EVP_CIPHER_CTX_FLAG_WRAP_ALLOW` flag).

sfackler commented on openssl/src/cipher_ctx.rs:
```diff
@@ -509,6 +509,14 @@ impl CipherCtxRef {
 
+    /// Set ctx flags.
+    #[corresponds(EVP_CIPHER_CTX_set_flags)]
+    pub fn set_flags(&mut self, flags: i32) {
```
If this is going to be publicly exposed, it should take a strongly-typed bitflags value.

mingweishih replied:
Changed to private, which is sufficient for this PR.

mingweishih replied:
I made the function public and make the flags as bitflags vlaue.

sfackler commented on openssl/src/symm.rs:
```diff
@@ -503,6 +539,11 @@ impl Crypter {
 
+        if cfg!(ossl102) {
+            // Explicitly opt in the AES key wrap feature
+            ctx.set_flags(ffi::EVP_CIPHER_CTX_FLAG_WRAP_ALLOW);
```
I don't think this should be done automatically.

mingweishih replied:
I updated to only set the flag if it's an aes wrap `cipher`. Does this work for you?

sfackler replied:
I don't think this should be done automatically.

mingweishih replied:
Reverted the change in symm.

mingweishih commented:
I addressed the comments. Would the current change work for you? I need this feature for a project. Hopefully it can go in for the upcoming release. Thanks!

sfackler commented:
Thanks!
